src/helpers/messaging.rs

```diff
@@ -182,16 +267,18 @@ impl Gateway {
                 ::tokio::select! {
                     Some(receive_request) = recv_rx.recv() => {
                         tracing::trace!("new {:?}", receive_request);
-                        receive_buf.receive_request(receive_request.channel_id, receive_request.record_id, receive_request.sender);
+                        let total_records = *channels.lock().unwrap().get(&receive_request.channel_id.step).expect("unknown channel");
+                        receive_buf.receive_request(receive_request.channel_id, receive_request.record_id, receive_request.sender, total_records);
                     }
                     Some((channel_id, messages)) = message_stream.next() => {
                         tracing::trace!("received {} bytes from {:?}", messages.len(), channel_id);
                         receive_buf.receive_messages(&channel_id, &messages);
                     }
                     Some((channel_id, envelope)) = send_rx.recv(), if pending_sends.is_empty() => {
```

Comment:
This arrangement means that we need to do a hashtable lookup for every send.  We put the channel id, record number, message, and max record number in every message we send.

I've been trying to work out how I would do this for the ordering_mpsc I put together, and I don't think that this is the way.  The nice thing there is that the channel only carries a record ID and message, with the other stuff being setup at creation time.  As much as possible, I'd like to get to the point where we don't replicate information.  (It's necessary here because we have a single channel for all steps, but I think that's something we want to get away from.)

I think that we can - in most cases - know how many records there will be before we create any such channel.  That simplifies things because the channel can be constructed with knowledge of when it ends.

The indeterminate length version is a harder problem.  For that, it might be necessary to have a different implementation that uses an atomic to communicate where to end.  That is slightly more expensive to read, but it should be possible to set that to a non-zero value once the last record is sent.  It should be possible to deal with that by presenting different functions for each.

Reply:
Agree, I don't want this to be the final implementation.

And I agree it's knowable (in most cases) how many records there will be before the channel is first used, but since our current channel construction is "create on first use", will require some restructuring of the code to actually make use of that knowledge.